File: cast-extractor/src/main/java/com/libreshockwave/tools/preview/SoundPreview.java
package com.libreshockwave.tools.preview;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.chunks.SoundChunk;
import com.libreshockwave.tools.model.CastMemberInfo;
import com.libreshockwave.tools.model.FrameAppearance;
import com.libreshockwave.tools.scanning.MemberResolver;
import com.libreshockwave.tools.score.FrameAppearanceFinder;

import java.util.List;

/**
 * Generates sound preview content.
 */
public class SoundPreview {

    private final FrameAppearanceFinder appearanceFinder = new FrameAppearanceFinder();

    /**
     * Formats sound details for display.
     */
    public String format(DirectorFile dirFile, CastMemberInfo memberInfo) {
        SoundChunk soundChunk = MemberResolver.findSoundForMember(dirFile, memberInfo.member());

        StringBuilder sb = new StringBuilder();
        sb.append("=== SOUND: ").append(memberInfo.name()).append(" ===\n\n");
        sb.append("Member ID: ").append(memberInfo.memberNum()).append("\n\n");

        if (soundChunk != null) {
            sb.append("--- Audio Properties ---\n");
            sb.append("Codec: ").append(soundChunk.isMp3() ? "MP3" : "PCM (16-bit)").append("\n");
            sb.append("Sample Rate: ").append(soundChunk.sampleRate()).append(" Hz\n");
            sb.append("Bits Per Sample: ").append(soundChunk.bitsPerSample()).append("\n");
            sb.append("Channels: ").append(soundChunk.channelCount() == 1 ? "Mono" : "Stereo").append("\n");
            sb.append("Duration: ").append(String.format("%.2f", soundChunk.durationSeconds())).append(" seconds\n");
            sb.append("Audio Data Size: ").append(soundChunk.audioData().length).append(" bytes\n");
        } else {
            sb.append("[Sound data not found]\n");
        }

        // Show frame appearances from score
        sb.append("\n--- Score Appearances ---\n");
        List<FrameAppearance> appearances = appearanceFinder.find(dirFile, memberInfo.memberNum());
        if (appearances.isEmpty()) {
            sb.append("Not used in score\n");
        } else {
            sb.append(appearanceFinder.format(appearances)).append("\n");
            // Show detailed list if not too many
            if (appearances.size() <= 20) {
                sb.append("\nDetailed appearances:\n");
                for (FrameAppearance app : appearances) {
                    sb.append(String.format("  Frame %d, %s", app.frame(), app.channelName()));
                    if (app.frameLabel() != null) {
                        sb.append(" [").append(app.frameLabel()).append("]");
                    }
                    sb.append("\n");
                }
            }
        }

        return sb.toString();
    }

    /**
     * Returns true if the sound chunk exists and is playable.
     */
    public boolean isPlayable(DirectorFile dirFile, CastMemberInfo memberInfo) {
        SoundChunk soundChunk = MemberResolver.findSoundForMember(dirFile, memberInfo.member());
        return soundChunk != null;
    }
}

File: cpp/src/editor/preview/SoundPreview.cpp
#include "libreshockwave/editor/preview/SoundPreview.hpp"

#include <iomanip>
#include <sstream>
#include <string>

#include "libreshockwave/chunks/SoundChunk.hpp"
#include "libreshockwave/editor/preview/PreviewFormatUtils.hpp"
#include "libreshockwave/editor/scanning/MemberResolver.hpp"
#include "libreshockwave/editor/score/FrameAppearanceFinder.hpp"

namespace libreshockwave::editor::preview {
namespace {

std::string formatSeconds(double value) {
    std::ostringstream out;
    out << std::fixed << std::setprecision(2) << value;
    return out.str();
}

} // namespace

std::string SoundPreview::format(DirectorFile& dirFile, const model::CastMemberInfo& memberInfo) const {
    const auto sound = scanning::MemberResolver::findSoundForMember(dirFile, memberInfo.member);

    std::string out;
    PreviewFormatUtils::appendMemberHeader(out, "SOUND", memberInfo, true);
    if (sound != nullptr) {
        out += "--- Audio Properties ---\n";
        out += "Codec: " + std::string(sound->isMp3() ? "MP3" : "PCM (16-bit)") + "\n";
        out += "Sample Rate: " + std::to_string(sound->sampleRate()) + " Hz\n";
        out += "Bits Per Sample: " + std::to_string(sound->bitsPerSample()) + "\n";
        out += "Channels: " + std::string(sound->channelCount() == 1 ? "Mono" : "Stereo") + "\n";
        out += "Duration: " + formatSeconds(sound->durationSeconds()) + " seconds\n";
        out += "Audio Data Size: " + std::to_string(sound->audioData().size()) + " bytes\n";
    } else {
        out += "[Sound data not found]\n";
    }

    out += "\n--- Score Appearances ---\n";
    const score::FrameAppearanceFinder appearanceFinder;
    const auto appearances = appearanceFinder.find(dirFile, memberInfo.memberNum);
    PreviewFormatUtils::appendScoreAppearances(out, appearances, appearanceFinder, false);
    return out;
}

bool SoundPreview::isPlayable(DirectorFile& dirFile, const model::CastMemberInfo& memberInfo) const {
    return scanning::MemberResolver::findSoundForMember(dirFile, memberInfo.member) != nullptr;
}

} // namespace libreshockwave::editor::preview
